wandb/sdk/wandb_run.py

```diff
@@ -2957,6 +2957,37 @@ def finish_artifact(
             finalize=True,
         )
 
+    def _download_artifact(
+        self,
+        artifact: Artifact,
+        root: Optional[str] = None,
+        recursive: bool = False,
+        allow_missing_references: bool = False,
+    ) -> FilePathStr:
+        if self._backend and self._backend.interface:
+            if not self._settings._offline:
+                future = self._backend.interface.download_artifact(
+                    artifact.qualified_name,
+                    root,
+                    recursive,
+                    allow_missing_references,
+                )
+                if hasattr(future, "response"):
+                    if future.response.download_artifact_response.error_message:
+                        raise ValueError(
+                            f"Error downloading artifact: {future.response.download_artifact_response.error_message}"
+                        )
+                    download_path = (
+                        future.response.download_artifact_response.file_download_path
+                    )
+                    return FilePathStr(download_path)
+                return FilePathStr("")
+        return artifact._download(
```

Comment:
This is cheating, we won't be able to do this with non-Python background process implementations (Nexus). Let's raise an exception in this case. This will be a breaking change though...

Reply:
Since we're leaving the reference downloads upto the python user process, can the offline case not be included there too?

Reply:
This is a temporary solution until we finish the Nexus implementation. Eventually we want to remove the download implementation from Python.

Reply:
Noted. I changed this bit of code - made it so that we raise an error if the user tries to download while in `offline` mode